### PR TITLE
Handle methods with an obvious negation in the non-minimal bool lint

### PR DESCRIPTION
**Pull request by LaurentMazare**
As suggested in #1930 warn on negations of `is_some`, `is_none`, `is_ok`, or `is_err`.

This is done using a const array to store methods with obvious negations. Not sure if this is the idiomatic way to do it, at least it should be reasonably efficient as there are only four elements involved.

### CODE REVIEW

File: clippy_lints/src/booleans.rs [hunk 1]
```diff
@@ -44,6 +44,13 @@ declare_lint! {
     "boolean expressions that contain terminals which can be eliminated"
 }
 
+const METHODS_WITH_NEGATION: [(&str, &str); 4] = [
```
Comment (oli-obk) [inline clippy_lints/src/booleans.rs]: What do you think about mentioning each pair only once and iterating over the list twice?

File: clippy_lints/src/booleans.rs [hunk 2]
```diff
@@ -406,6 +437,7 @@ impl<'a, 'tcx> Visitor<'tcx> for NonminimalBoolVisitor<'a, 'tcx> {
         match e.node {
             ExprBinary(binop, _, _) if binop.node == BiOr || binop.node == BiAnd => self.bool_expr(e),
             ExprUnary(UnNot, ref inner) => if self.cx.tables.node_types()[inner.hir_id].is_bool() {
+                self.handle_method_call_in_not(e, inner);
```
Comment (oli-obk) [inline clippy_lints/src/booleans.rs]: I would prefer to include this as part of https://github.com/rust-lang-nursery/rust-clippy/pull/2216/files#diff-2874c632e14921250b07b4805c090f19R181 so things like `!(foo.is_some() && !bar)` gets optimized to `foo.is_none() || bar`

File: clippy_lints/src/booleans.rs (hunk 1 above)
Comment (LaurentMazare) [reply clippy_lints/src/booleans.rs]: Seems better indeed, changed.

File: clippy_lints/src/booleans.rs (hunk 2 above)
Comment (LaurentMazare) [reply clippy_lints/src/booleans.rs]: I gave this a short try without much success. I'm not sure to understand the whole architecture correctly but wouldn't that require 'expr.simplify()' (in the bool_expr function) to also handle these negations ?

Comment (oli-obk) [reply clippy_lints/src/booleans.rs]: Hmmm since I wrote the code, I'll see if I can figure it out. Looks certainly nontrivial

Comment (LaurentMazare) [reply clippy_lints/src/booleans.rs]: Ok, let me know if there is anything I can do to help.

Comment (oli-obk) [reply clippy_lints/src/booleans.rs]: So. it looks like it's not necessary to modify the `simplify` method. Or I haven't been able to create horrible enough test cases that trigger said issue.

**Comment by LaurentMazare** [comment]
@oli-obk thanks this looks pretty simple indeed.
However if I understand correctly this does not produce any warning in the `if !a.is_some()` case but only warns on more complex expressions. Is it on purpose or do you think we should suggest switching to `is_none` in this case (or maybe I'm just misreading it) ?

**Comment by oli-obk** [comment]
Nope, that's exactly what's happening. Even for `!(foo == bar)` I think. Do you want to take the PR over again? Should be as simple as checking whether running the `suggest` function actually does anything. We might need to track that in some form, I don't think currently report whether any simplification has been done

**Comment by LaurentMazare** [comment]
I just pushed some changes along the lines of what you suggested, i.e. getting `suggest` to return whether some simplifications has been made.
It seems to work fine but overall I'm a bit worried about the current complexity of the `suggest` function, maybe I can have a look at if it's possible to simplify it in another pull request.

**Comment by oli-obk** [comment]
> I just pushed some changes along the lines of what you suggested, i.e. getting suggest to return whether some simplifications has been made.

lgtm, I'm just going through a rustup then I'll merge the PR

> maybe I can have a look at if it's possible to simplify it in another pull request.

That would be great

**Pull request merged**